testutils/utils.go

```diff
@@ -0,0 +1,399 @@
+	if !exists {
+		sr = nil
+	}
+
+	return sr, err
+}
+
+func DeleteServiceDefaults(ctx context.Context, k8sClient client.Client, service string) error {
+	sd := new(consulk8s.ServiceDefaults)
+	err := deleteK8sObject(ctx, k8sClient, service, sd)
+
+	return err
+}
+
+func DeleteConsulServiceRoute(ctx context.Context, k8sClient client.Client, name string) error {
+	csr := new(v1alpha1.ConsulServiceRoute)
+	err := deleteK8sObject(ctx, k8sClient, name, csr)
+
+	return err
+}
+
+func UpdateConsulServiceRoute(ctx context.Context, k8sClient client.Client, updated *v1alpha1.ConsulServiceRoute) error {
+	err := k8sClient.Update(ctx, updated)
+	if err != nil {
+		return err
+	}
+
+	err = waitForConsulServiceRouteToBeUpToDate(ctx, k8sClient, updated)
+
+	return err
+}
+
+func ExpectConsulServiceRoute(ctx context.Context, k8sClient client.Client, name string, expected consulk8s.ServiceRoute) {
+	csr, err := GetConsulServiceRoute(ctx, k8sClient, name)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+	gomega.Expect(csr).NotTo(gomega.BeNil())
+	gomega.Expect(csr.Finalizers).To(gomega.ContainElement(ServiceFinalizer))
+
+	gomega.Expect(csr.Status.UpdatedAt).NotTo(gomega.BeEmpty())
+	gomega.Expect(csr.Status.ContentSHA).To(gomega.Equal(getConsulServiceRouteContentSHA(csr)))
+
+	gomega.Expect(csr.Spec.Route).To(gomega.Equal(expected))
+}
+
+func CreateHTTPPathPrefixRoute(service, pathPrefix string) consulk8s.ServiceRoute {
+	result := consulk8s.ServiceRoute{
+		Match:       &consulk8s.ServiceRouteMatch{HTTP: &consulk8s.ServiceRouteHTTPMatch{PathPrefix: pathPrefix}},
+		Destination: &consulk8s.ServiceRouteDestination{Service: service},
+	}
+
+	return result
+}
+
+func CreateConsulServiceRoutes(ctx context.Context, k8sClient client.Client, serviceRouterName string, routes []consulk8s.ServiceRoute) {
+	for _, serviceRoute := range routes {
+		err := CreateConsulServiceRoute(ctx, k8sClient, serviceRouterName, serviceRoute)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}
+
+	err := WaitForServiceRouterToBeCreated(ctx, k8sClient, serviceRouterName)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+}
+
+func DeleteConsulServiceRoutes(ctx context.Context, k8sClient client.Client, serviceRouterName string, routes []consulk8s.ServiceRoute) {
+	for _, serviceRoute := range routes {
+		err := DeleteConsulServiceRoute(ctx, k8sClient, serviceRoute.Destination.Service)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}
+
+	serviceRouter, err := GetServiceRouter(ctx, k8sClient, serviceRouterName)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+	if serviceRouter == nil {
+		return
+	}
+
+	err = deleteK8sObject(ctx, k8sClient, serviceRouterName, serviceRouter)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+}
+
+func WaitForServiceRouterToBeCreated(ctx context.Context, k8sClient client.Client, name string) error {
+	serviceRouter := new(consulk8s.ServiceRouter)
+
+	for i := 0; i < 10; i++ {
+		time.Sleep(100 * time.Millisecond)
```

Comment:
You can extract the logic of the `for` loop along with `time.Sleep(100 * time.Millisecond)` to a separate function here in testutils and reuse it on all 4 occasions